core/corehttp/gateway_handler_unixfs_dir.go

```diff
@@ -108,36 +109,46 @@ func (i *gatewayHandler) serveDirectory(ctx context.Context, w http.ResponseWrit
 		return
 	}
 
+	// Optimization 1:
+	// List children without fetching their root blocks (fast, but no size info)
+	results, err := i.api.Unixfs().Ls(ctx, resolvedPath, options.Unixfs.ResolveChildren(false))
+	if err != nil {
+		internalWebError(w, err)
+		return
+	}
+
 	// storage for directory listing
-	var dirListing []directoryItem
-	dirit := dir.Entries()
-	for dirit.Next() {
-		size := "?"
-		if s, err := dirit.Node().Size(); err == nil {
-			// Size may not be defined/supported. Continue anyways.
-			size = humanize.Bytes(uint64(s))
-		}
+	dirListing := make([]directoryItem, 0, len(results))
 
-		resolved, err := i.api.ResolvePath(ctx, ipath.Join(resolvedPath, dirit.Name()))
-		if err != nil {
+	for link := range results {
+		if link.Err != nil {
 			internalWebError(w, err)
 			return
 		}
-		hash := resolved.Cid().String()
-
-		// See comment above where originalUrlPath is declared.
+		hash := link.Cid.String()
 		di := directoryItem{
-			Size:      size,
-			Name:      dirit.Name(),
-			Path:      gopath.Join(originalUrlPath, dirit.Name()),
+			Size:      "", // no size because we did not fetch child nodes
```

Comment:
For consistency with default size value set below for items we fetch the size for?

```suggestion
			Size:      "?", // unknown size because we did not fetch child nodes or node does not support size
```

Reply:
I removed "?" on purpose: a wall of "?" does not improve UX in any way. 

The "?" makes sense only when dag-pb directory points at non-unixfs CID that does not have size info.

Reply:
Why are we fetching sizes in the first place ?

It costy and that info I personally don't care. Can't just nuke this code ? Do some people care about that ?

Reply:
Usually it is useless data, but sometimes size column allows user to reason how long the download will take: https://dist.ipfs.io/go-ipfs/v0.12.2

Personally, I'd like to stop fetching sizes by default too, but we should add an opt-in way for fetching them, if user really cares. In my mind that is something we could discuss after this non-invasive PR is merged.